### PR TITLE
Blockstore should drop signals before validator exit

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

blockstore holds two set of signal senders, and blockstore is Arc in many places, including RpcSubscription. During shutting down, a deadlock may occurs, when the signal receiver is using a blocking call and the blockstore ref count never goes to 0. 

#### Summary of Changes

- Add drop_signals function on blockstore. When exiting, drop the signals to avoid deadlock.
- Change signals to be protected by mutex. 

Fixes #

### CODE REVIEW

File: ledger/src/blockstore.rs [hunk 1]
```diff
@@ -444,8 +444,8 @@ impl Blockstore {
             block_height_cf,
             program_costs_cf,
             bank_hash_cf,
-            new_shreds_signals: vec![],
-            completed_slots_senders: vec![],
+            new_shreds_signals: Mutex::<Vec<Sender<bool>>>::default(),
```
Comment (jeffwashington) [inline ledger/src/blockstore.rs:447]: I think you can say `Mutex::default()` here.

**Comment by jeffwashington** [comment]
is this solving test hanging? I don't recall the context.
Does `RwLock` work better here or is `Mutex` appropriate? I'm not familiar with the usage patterns in the normal validator.

**Comment by HaoranYi** [comment]
> is this solving test hanging? I don't recall the context. Does `RwLock` work better here or is `Mutex` appropriate? I'm not familiar with the usage patterns in the normal validator.

Yes. This one is a follow up of the https://github.com/solana-labs/solana/pull/24007. 

In https://github.com/solana-labs/solana/pull/24007, we fix the hang by using nonblocking call inside RpcCompletedSlotService. 

This change implement the idea suggested by @sakridge - when shutting down, we close all the signal senders in blockstore, so that all receivers whether blocking or nonblocking will return on receive. 

In this case, I don't think RwLock and Mutex will make much difference. The vec inside the lock is only written twice, once at the initialization, and once cleared at the shutdown. And there is only one thread reading it inside blockstore.

**Comment by carllin** [comment]
Hmm, is this going to fix anything? I feel like the `exit` signal should be sufficient for exit.

The signals are created here: https://github.com/solana-labs/solana/blob/51b37f0184b15c02306d3047770d00755ab9c764/core/src/validator.rs#L495-L496

The consumers are:

1. RPC https://github.com/solana-labs/solana/blob/51b37f0184b15c02306d3047770d00755ab9c764/core/src/validator.rs#L875-L879 which also gets an exit signal
2.  The tvu: https://github.com/solana-labs/solana/blob/51b37f0184b15c02306d3047770d00755ab9c764/core/src/validator.rs#L895, which also gets an exit signal

Thus it seems like setting the `exit` flag from the validator exit: https://github.com/solana-labs/solana/blob/c6ef522118108cc951e2f1459566647a2ff05f14/core/src/validator.rs#L1006 should be sufficient

**Comment by HaoranYi** [comment]
@carllin 
You are right. This one doesn't fix anything since https://github.com/solana-labs/solana/pull/24007 already have fixed the hang. The rpc code you listed above is recent change https://github.com/solana-labs/solana/pull/24007,  that is used to fix the hang on the blocking call. 

This one can be thought of as more of refactor and protection. 
- We make the signal vec the blockstore private and expose API for access them
- We clear the signal vec before shutdown, which might help to speedup the shutdown (no need to wait till timeout)

Also, while refactoring this, I notice another potential issue https://github.com/solana-labs/solana/pull/24051
We could potentially missing waking up signals in replaying stage. Please help to take a look at this as well. Let me know what do you think. Thanks!

**Review by carllin — APPROVED** [review]
@HaoranYi understood, thanks for explaining!

**Pull request merged**